### DISCUSSION TITLE
Frosting: How to get a behavior similar to ".WithCriteria()"?

### DISCUSSION THREAD

**Discussion by bergziege** [cake-build/cake#2951] [Q&A] [answered]
Hi, just switched from cake script to Frosting and run into an annoying problem.
When using cake script you could append `.WithCriteria` to a task definition and that said task only executes when the criteria matches. Therefore said task would not be listet in the console output until the criteria matches.

Now I am trying to get a similar behavior with frosting. I understand that it is not possible with attributes as those need to be compile time constant.
I am currently using if/else to check criterias and execute task content. Which has the side effect that the task gets reported in the console even its not really doing anything besides confusing the user ;-)

Is there another way (via the context?) to skip a task when certain criterias are not fulfilled or at least hide it from console output?

**Comment by patriksvensson** [comment, marked as answer]
@bergziege You can override the `ShouldRun` method in the task to accomplish this.

```csharp
[Dependency(typeof(Build))]
public sealed class MyTask : FrostingTask<Context>
{
    public override bool ShouldRun(Context context)
    {
        // Criteria here
        return true;
    }

    public override void Run(Context context)
    {
    }
}
```

**Reply by pascalberger** [reply]
This has now also been documented [here](https://cakebuild.net/docs/writing-builds/tasks/criteria#cake-frosting)

**Reply by arturcic** [reply]
@patriksvensson, @pascalberger  What if I want to have a message displayed if the criteria is not fulfilled. Similar to the  `.WithCriteria(criteria, message)`.

What if we want to have separate message for each criteria in case we have more than one criteria?

**Reply by pascalberger** [reply]
Defining a message for skipped criterias is currently not supported in Frosting

**Reply by WebDucer** [reply]
We workaround this by writing the "message" manually, like this:

```csharp
public override bool ShouldRun(BuildContext context)
{
  var apiToken = context.EnvironmentVariable(EnvNames.NUGET_API_TOKEN);
  if (string.IsNullOrWhiteSpace(apiToken))
  {
    context.Information($"SKIPPED {nameof(PublishTask)}: NuGet API token not set");
    return false;
  }

  if (!context.GetFiles(context.BuildConfig.PackageDirectory.FullPath + "/*.nupkg").Any())
  {
    context.Information($"SKIPPED {nameof(PublishTask)}: No NuGet packages found in '{context.BuildConfig.PackageDirectory}'");
    return false;
  }
  return true;
}
```

Then you will get the following output:

```
----------------------------------------
Setup
----------------------------------------

========================================
Show-Info
========================================
Build configuration
┌────────────────────┬──────────────────────────────────────────┐
│ ProjectName        │ MY-PROJECT                               │
│ BuildConfiguration │ Release                                  │
│ SourceDirectory    │ C:/src/Libs/MY-PROJECT/src               │
│ OutputDirectory    │ C:/src/Libs/MY-PROJECT/Artifacts         │
│ SemanticVersion    │ 0.8.0-beta.10                            │
│ BuildDate          │ 2021-04-28 10:08                         │
│ Sha                │ 8d0121f9953f0b8c1454133ec6084f5be8d1b3e5 │
│ PackageDirectory   │ C:/src/Libs/MY-PROJECT/Artifacts/NuGet   │
│ NuGetVersion       │ 0.8.0-beta0010                           │
└────────────────────┴──────────────────────────────────────────┘
SKIPPED PublishTask: NuGet API token not set

========================================
Publish
========================================

----------------------------------------
Teardown
----------------------------------------

Task                          Duration
--------------------------------------------------
Setup                         00:00:01.7292003
Show-Info                     00:00:00.1331840
NuGet-Publish                 Skipped
Teardown                      00:00:00.0019460
--------------------------------------------------
Total:                        00:00:01.8653543
```

**Comment by patriksvensson** [comment]
I've created an issue for this here: https://github.com/cake-build/cake/issues/3328
